lib/config/formatter.go

```diff
@@ -0,0 +1,374 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"regexp"
+	"runtime"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+type textFormatter struct {
+	// ComponentPadding is a padding to pick when displaying
+	// and formatting component field, defaults to DefaultComponentPadding
+	ComponentPadding int
+	// EnableColors enables colored output
+	EnableColors bool
+	// FormatCaller is a function to return (part) of source file path for output.
+	// Defaults to filePathAndLine() if unspecified
+	FormatCaller func() (caller string)
+	// InputFormat is the parsed format of each log line
+	LogFormat []string
+	// TimestampEnabled specifies if timestamp is enabled in logs
+	timestampEnabled bool
+	// CallerEnabled specifies if caller is enabled in logs
+	callerEnabled bool
+}
+
+type writer struct {
+	bytes.Buffer
+}
+
+const (
+	noColor        = -1
+	red            = 31
+	yellow         = 33
+	blue           = 36
+	gray           = 37
+	levelField     = "level"
+	componentField = "component"
+	callerField    = "caller"
+	timestampField = "timestamp"
+)
+
+// CheckAndSetDefaults checks and sets log format configuration
+func (tf *textFormatter) CheckAndSetDefaults() error {
+	// set padding
+	if tf.ComponentPadding == 0 {
+		tf.ComponentPadding = trace.DefaultComponentPadding
+	}
+	// set caller
+	tf.FormatCaller = formatCallerWithPathAndLine
+
+	// set log formatting
+	if tf.LogFormat == nil {
+		tf.timestampEnabled = true
+		tf.callerEnabled = true
+		tf.LogFormat = KnownFormatFields.names()
+		return nil
+	}
+	// parse input
+	res, err := parseInputFormat(tf.LogFormat)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if contains(res, timestampField) {
```

Comment:
Sorry for late reply: `contains` is a dupe for [api.types.SliceContainsStr](https://github.com/gravitational/teleport/blob/4d916e4bfef0c5722048679cf792cc24d4bf1298/api/utils/slices.go#L55)